boat_simulator/nodes/data_collection/data_collection_node.py

```diff
@@ -0,0 +1,222 @@
+            namespace="",
+            parameters=[
+                ("file_name", rclpy.Parameter.Type.STRING),
+                ("qos_depth", rclpy.Parameter.Type.INTEGER),
+                ("topics", rclpy.Parameter.Type.STRING_ARRAY),
+                ("bag", rclpy.Parameter.Type.BOOL),
+                ("json", rclpy.Parameter.Type.BOOL),
+                ("write_period_sec", rclpy.Parameter.Type.DOUBLE),
+            ],
+        ),
+        all_parameters = self._parameters
+        for name, parameter in all_parameters.items():
+            value_str = str(parameter.value)
+            self.get_logger().debug(f"Got parameter {name} with value {value_str}")
+
+    def __init_storage_files(self):
+        """Initializes storage files for data logging."""
+        self.get_logger().debug("Initializing storrage files...")
+
+        if self.use_json:
+            self.__data_to_write = {}
+            self.__json_index_counter = 0
+            self.__start_time = self.get_clock().now().nanoseconds
+            json_file_path = "./" + self.file_name + ".json"
+            if os.path.exists(json_file_path):
+                self.get_logger().warn(
+                    f"JSON file with name {self.file_name} already exists. Overriding old file..."
+                )
+                os.remove(json_file_path)
+            self.__json_file = open(json_file_path, "a")
+            self.__json_file.write("[\n")  # Open JSON array
+
+        if self.use_bag:
+            self.__writer = rosbag2_py.SequentialWriter()
+            storage_options = rosbag2_py._storage.StorageOptions(
+                uri=self.file_name, storage_id="sqlite3"
+            )
+            converter_options = rosbag2_py._storage.ConverterOptions("", "")
+            self.__writer.open(storage_options, converter_options)
+
+    def __init_msg_types_dict(self):
+        """Prepare dictionary of all msg types with key name and value class"""
+        self.__msg_types_dict = {}
+        for name, cls in inspect.getmembers(custom_interfaces.msg, inspect.isclass):
+            if not name.startswith("_"):
+                self.__msg_types_dict[name] = cls
+
+    def __init_subscriptions(self):
+        """Initialize the subscriptions for this node. These subscriptions pertain to the topics
+        from which data will be collected."""
+        self.get_logger().debug("Initializing subscriptions...")
+
+        def create_subscription_with_callback(topic_name, msg_type):
+            return self.create_subscription(
+                msg_type=msg_type,
+                topic=topic_name,
+                callback=lambda msg: self.__general_sub_callback(msg, topic_name),
+                qos_profile=self.get_parameter("qos_depth").get_parameter_value().integer_value,
+            )
+
+        self.__record_topics_sub = []
+
+        # Assuming topics are specified as ['topic_name', 'topic_type']
+        for i in range(1, len(self.topics_to_record), 2):
```

Comment:
This is kinda nitpicky, but complex for loops like this can be hard to read. I have a suggestion to make it more readable, so let me know what you think:

```python
# Get topic names by extracting from all evenly indexed elements and all topic types from oddly
# indexed elements since it is assumed that topic name and type alternate in the list
# For example, [name1, type1, name2, type2, ...]
topic_names = self.topics_to_record[::2]
topic_types = self.topics_to_record[1::2]
for topic_name, msg_type_name in zip(topic_names, topic_types):
    # Do the for loop stuff ...
```